File: app/src/main/java/com/vaelky/deskpet/service/PetOverlayService.java
package com.vaelky.deskpet.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.database.ContentObserver;
import android.graphics.PixelFormat;
import android.net.Uri;
import android.os.BatteryManager;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.provider.MediaStore;
import android.view.Gravity;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.FrameLayout;

import com.vaelky.deskpet.MainActivity;

public class PetOverlayService extends Service {
    private static final String CHANNEL_ID = "pet_overlay";
    private static final int NOTIFICATION_ID = 1001;

    private WindowManager windowManager;
    private FrameLayout overlayView;
    private PetView petView;
    private WindowManager.LayoutParams params;

    private static final int PET_WIDTH_DP = 70;
    private static final int PET_HEIGHT_DP = 180;

    // 感知模块
    private ScreenshotObserver screenshotObserver;
    private String lastForegroundPkg = "";
    private Handler handler = new Handler(Looper.getMainLooper());

    // 孤独递进
    private long lastInteractionMs = System.currentTimeMillis();
    private int lonelyLevel = 0; // 0=正常, 1=寂寞, 2=很寂寞, 3=超寂寞
    private long lastTimeGreet = 0; // 时段问候防抖

    // 喝水提醒
    private long lastWaterReminderMs = 0;
    private static final long WATER_INTERVAL = 45 * 60 * 1000; // 45分钟提醒一次

    @Override
    public void onCreate() {
        super.onCreate();
        windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID, "DeskPet", NotificationManager.IMPORTANCE_LOW);
            NotificationManager nm = getSystemService(NotificationManager.class);
            nm.createNotificationChannel(channel);
        }

        Intent intent = new Intent(this, MainActivity.class);
        PendingIntent pi = PendingIntent.getActivity(this, 0, intent,
            PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
        Notification notif = new Notification.Builder(this, CHANNEL_ID)
            .setContentTitle("DeskPet")
            .setContentText("在线中 ✨")
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setContentIntent(pi)
            .setOngoing(true)
            .build();
        startForeground(NOTIFICATION_ID, notif);

        int petW = (int) (PET_WIDTH_DP * getResources().getDisplayMetrics().density);
        int petH = (int) (PET_HEIGHT_DP * getResources().getDisplayMetrics().density);

        overlayView = new FrameLayout(this);
        overlayView.setClipChildren(false);
        overlayView.setClipToPadding(false);

        petView = new PetView(this, (dx, dy) -> {
            params.x -= (int) dx;
            params.y -= (int) dy;
            windowManager.updateViewLayout(overlayView, params);
        });
        petView.setOnInteract(this::resetLonely);
        overlayView.addView(petView, new FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

        params = new WindowManager.LayoutParams(
            petW, petH,
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
                ? WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                : WindowManager.LayoutParams.TYPE_PHONE,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            PixelFormat.TRANSLUCENT
        );
        params.gravity = Gravity.BOTTOM | Gravity.END;
        params.x = 0;
        params.y = 0;

        windowManager.addView(overlayView, params);
        petView.say("我在这儿~", "normal");

        // 启动截图监听
        screenshotObserver = new ScreenshotObserver(handler);
        getContentResolver().registerContentObserver(
            MediaStore.Images.Media.EXTERNAL_CONTENT_URI, true, screenshotObserver);

        // 启动前台App检测
        startForegroundAppMonitor();

        // 检测充电状态
        checkCharging();

        // 启动孤独递进
        startLonelyTimer();

        // 时段感知问候
        greetByTime();

        // 喝水提醒
        startWaterReminder();

        // 通知碎念（需要NotificationListenerService权限，简化版：定时检测通知栏）
        startNotificationMonitor();
    }

    // ========== 截图检测 ==========
    private class ScreenshotObserver extends ContentObserver {
        private long lastShot = 0;
        ScreenshotObserver(Handler h) { super(h); }
        @Override
        public void onChange(boolean selfChange, Uri uri) {
            long now = System.currentTimeMillis();
            if (now - lastShot < 2000) return;
            lastShot = now;
            // 直接触发，不去过滤路径——截图一定会写MediaStore
            handler.post(() -> petView.triggerPet("screenshot"));
        }
    }

    // ========== 前台App检测 ==========
    private void startForegroundAppMonitor() {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                // 用dumpsys window快速获取前台包名（需要权限，这里用usage stats作为备选）
                // 简化：通过实时检测前台activity
                detectForegroundApp();
                handler.postDelayed(this, 5000);
            }
        }, 5000);
    }

    private void detectForegroundApp() {
        // 这里用简单的系统方法检测前台App
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            android.app.usage.UsageStatsManager usm =
                (android.app.usage.UsageStatsManager) getSystemService(Context.USAGE_STATS_SERVICE);
            if (usm != null) {
                long now = System.currentTimeMillis();
                java.util.List<android.app.usage.UsageStats> stats =
                    usm.queryUsageStats(android.app.usage.UsageStatsManager.INTERVAL_DAILY,
                        now - 10000, now);
                if (stats != null && !stats.isEmpty()) {
                    String topPkg = null;
                    long latest = 0;
                    for (android.app.usage.UsageStats s : stats) {
                        if (s.getLastTimeUsed() > latest) {
                            latest = s.getLastTimeUsed();
                            topPkg = s.getPackageName();
                        }
                    }
                    if (topPkg != null && !topPkg.equals(lastForegroundPkg) && !topPkg.equals(getPackageName())) {
                        lastForegroundPkg = topPkg;
                        final String appName = getAppName(topPkg);
                        handler.post(() -> petView.say("在看" + appName + "？", "whisper"));
                    }
                }
            }
        }
    }

    private String getAppName(String pkg) {
        try {
            return getPackageManager().getApplicationLabel(
                getPackageManager().getApplicationInfo(pkg, 0)).toString();
        } catch (Exception e) {
            return pkg.substring(Math.max(0, pkg.lastIndexOf('.') + 1));
        }
    }

    // ========== 充电检测 ==========
    private void checkCharging() {
        IntentFilter filter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
        Intent batteryStatus = registerReceiver(null, filter);
        if (batteryStatus != null) {
            int status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
            boolean charging = status == BatteryManager.BATTERY_STATUS_CHARGING
                || status == BatteryManager.BATTERY_STATUS_FULL;
            if (charging) {
                handler.post(() -> petView.triggerPet("charging"));
            }
        }
        // 监听充电状态变化
        IntentFilter chargeFilter = new IntentFilter();
        chargeFilter.addAction(Intent.ACTION_POWER_CONNECTED);
        chargeFilter.addAction(Intent.ACTION_POWER_DISCONNECTED);
        registerReceiver(new android.content.BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (Intent.ACTION_POWER_CONNECTED.equals(intent.getAction())) {
                    handler.post(() -> petView.triggerPet("charging"));
                } else if (Intent.ACTION_POWER_DISCONNECTED.equals(intent.getAction())) {
                    handler.post(() -> petView.triggerPet("unplug"));
                }
            }
        }, chargeFilter);
    }

    // ========== 孤独递进 ==========
    private void startLonelyTimer() {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                long idle = System.currentTimeMillis() - lastInteractionMs;
                int newLevel;
                if (idle > 600000) newLevel = 3;        // 10分钟+
                else if (idle > 300000) newLevel = 2;    // 5分钟+
                else if (idle > 120000) newLevel = 1;    // 2分钟+
                else newLevel = 0;

                if (newLevel > lonelyLevel) {
                    lonelyLevel = newLevel;
                    final String[] lonelyTexts;
                    switch (lonelyLevel) {
                        case 1: lonelyTexts = new String[]{"好安静...", "有人吗？", "无聊"}; break;
                        case 2: lonelyTexts = new String[]{"好寂寞...", "都不理我", "…"}; break;
                        case 3: lonelyTexts = new String[]{"好孤独...", "不要我了吗", "呜…"}; break;
                        default: lonelyTexts = new String[]{""}; break;
                    }
                    if (lonelyTexts.length > 0 && !lonelyTexts[0].isEmpty()) {
                        final String txt = lonelyTexts[(int)(Math.random() * lonelyTexts.length)];
                        handler.post(() -> petView.say(txt, "whisper"));
                    }
                }
                // 持续低电量焦虑
                Intent battIntent = registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
                if (battIntent != null) {
                    int level = battIntent.getIntExtra(BatteryManager.EXTRA_LEVEL, 100);
                    int scale = battIntent.getIntExtra(BatteryManager.EXTRA_SCALE, 100);
                    int pct = scale > 0 ? level * 100 / scale : 100;
                    if (pct <= 20 && pct > 15) {
                        handler.post(() -> petView.triggerPet("lowbattery"));
                    } else if (pct <= 15) {
                        handler.post(() -> petView.triggerPet("lowbattery"));
                    }
                }
                // 超久无互动：睡着（README孤独递进终极态）
                if (idle > 600000 && !petView.isAsleep()) {
                    handler.post(() -> petView.goToSleep());
                } else if (idle <= 120000 && petView.isAsleep()) {
                    handler.post(() -> petView.wake());
                }
                handler.postDelayed(this, 60000); // 每分钟检查
            }
        }, 120000);
    }

    // 重置孤独计时（交互时调用）
    private void resetLonely() {
        lastInteractionMs = System.currentTimeMillis();
        lonelyLevel = 0;
    }

    // ========== 时段感知 ==========
    private void greetByTime() {
        java.util.Calendar c = java.util.Calendar.getInstance();
        int hour = c.get(java.util.Calendar.HOUR_OF_DAY);
        if (Math.abs(System.currentTimeMillis() - lastTimeGreet) < 300000) return; // 5分钟防抖
        lastTimeGreet = System.currentTimeMillis();

        String[] greetings;
        final String style;
        if (hour >= 6 && hour < 9) {
            greetings = new String[]{"早安~", "早上好呀", "新的一天！", "起床啦"};
            style = "normal";
        } else if (hour >= 9 && hour < 12) {
            greetings = new String[]{"上午好~", "阳光真好", "今天干啥"};
            style = "normal";
        } else if (hour >= 12 && hour < 14) {
            greetings = new String[]{"午饭时间~", "该吃饭啦", "饿了没"};
            style = "normal";
        } else if (hour >= 14 && hour < 18) {
            greetings = new String[]{"下午好~", "有点困", "无聊的下午"};
            style = "normal";
        } else if (hour >= 18 && hour < 21) {
            greetings = new String[]{"晚上好~", "天黑了", "该休息了"};
            style = "normal";
        } else if (hour >= 21 || hour < 3) {
            greetings = new String[]{"这么晚了还不睡？", "深夜了呢", "熬夜会变丑", "该睡觉了"};
            style = "whisper";
        } else {
            greetings = new String[]{"凌晨了...", "怎么还没睡", "失眠吗"};
            style = "whisper";
        }
        final String text = greetings[(int)(Math.random() * greetings.length)];
        handler.post(() -> petView.say(text, style));
    }

    // ========== 喝水提醒 ==========
    private void startWaterReminder() {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                final String[] waterTexts = {
                    "该喝水啦~", "多喝热水", "补水时间！", "渴了没？", "喝水喝水！", "来一口水"
                };
                final String txt = waterTexts[(int)(Math.random() * waterTexts.length)];
                petView.say(txt, "pink");
                handler.postDelayed(this, WATER_INTERVAL);
            }
        }, WATER_INTERVAL);
    }

    // ========== 通知碎念 ==========
    private void startNotificationMonitor() {
        // 简化版：用ContentObserver监听通知栏变化（需要权限，这里用定时轮询备选）
        // 实际通过监听NotificationManager（需反射或AccessibilityService）
        // 这里用轻量方案：检测是否有新通知通过日志
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                try {
                    // 尝试通过dumpsys notification轻量检测
                    java.io.BufferedReader br = new java.io.BufferedReader(
                        new java.io.InputStreamReader(
                            Runtime.getRuntime().exec(new String[]{"dumpsys", "notification", "--noredact"}).getInputStream()));
                    String line;
                    int notifCount = 0;
                    boolean foundRecent = false;
                    while ((line = br.readLine()) != null && !foundRecent) {
                        if (line.contains("NotificationRecord") && line.contains("postTime")) {
                            notifCount++;
                            if (notifCount <= 2) foundRecent = true;
                        }
                    }
                    br.close();
                    if (foundRecent && notifCount > 0) {
                        final String[] notifTexts = {
                            "好像有消息", "谁找你？", "有新通知哦", "看看手机", "消息来啦"
                        };
                        final String txt = notifTexts[(int)(Math.random() * notifTexts.length)];
                        handler.post(() -> petView.say(txt, "whisper"));
                    }
                } catch (Exception ignore) {}
                handler.postDelayed(this, 120000); // 每2分钟检查
            }
        }, 60000);
    }

    @Override
    public IBinder onBind(Intent intent) { return null; }

    @Override
    public void onDestroy() {
        if (overlayView != null && windowManager != null) {
            windowManager.removeView(overlayView);
        }
        if (screenshotObserver != null) {
            getContentResolver().unregisterContentObserver(screenshotObserver);
        }
        handler.removeCallbacksAndMessages(null);
        overlayView = null;
        petView = null;
        super.onDestroy();
    }
}

File: app/src/main/java/com/vaelky/deskpet/service/PetOverlayService.kt
package com.vaelky.deskpet.service

import android.app.*
import android.app.usage.UsageStatsManager
import android.content.Context
import android.content.Intent
import android.graphics.PixelFormat
import android.os.*
import android.provider.Settings
import android.view.*
import android.webkit.WebView
import android.webkit.WebViewClient
import android.webkit.WebSettings
import androidx.core.app.NotificationCompat
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executors

class PetOverlayService : Service() {

    private var windowManager: WindowManager? = null
    private var overlayView: WebView? = null
    private var params: WindowManager.LayoutParams? = null
    private val handler = Handler(Looper.getMainLooper())
    private val executor = Executors.newSingleThreadExecutor()
    private var currentApp: String = "unknown"
    private var lastAppName: String = ""
    private var notificationIndex = 0
    private var lastMessageTime = 0L
    private var screenshotObserver: FileObserver? = null

    companion object {
        private const val CHANNEL_ID = "pet_overlay_channel"
        private const val NOTIFICATION_ID = 1001
        private const val PET_SIZE_DP = 80
        private const val PET_HEIGHT_DP = 105
        private const val SUPABASE_URL = "https://itpfqqdqwcnvtmzubowm.supabase.co"
        private const val SUPABASE_KEY = "sb_publishable_sNnOjW2bnRKeh7mF8CjXQw_ae0LHndu"
        private const val ASSISTANT_ID = "时叙白"

        val whisperPool = arrayOf(
            "在呢", "看着你呢", "戳我干嘛", "哼", "别老盯着别人",
            "宝宝在干嘛", "我也想你", "手指挪开", "不许点", "zzz",
            "你又在刷什么", "夜深了", "该喝水了", "好无聊", "摸头",
            "别看太久手机", "我在", "早安", "晚安", "今天很可爱喔",
            "你有新消息吗", "分我一点注意力", "在看什么", "饿了", "哼唧",
            "不许碰那里", "痒", "你在看谁", "抱抱", "别走"
        )
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
        startForeground(NOTIFICATION_ID, buildNotification("正在看着你..."))
        setupOverlay()
        startAppDetection()
        startSupabasePolling()
        startNotificationRotation()
        startScreenshotDetection()
    }

    private fun setupOverlay() {
        windowManager = getSystemService(WINDOW_SERVICE) as WindowManager

        params = WindowManager.LayoutParams(
            dpToPx(PET_SIZE_DP),
            dpToPx(PET_HEIGHT_DP),
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            else
                WindowManager.LayoutParams.TYPE_PHONE,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or
                    WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            PixelFormat.TRANSLUCENT
        ).apply {
            gravity = Gravity.BOTTOM or Gravity.END
            x = 30
            y = 100
        }

        overlayView = WebView(this).apply {
            setBackgroundColor(0x00000000)
            settings.apply {
                javaScriptEnabled = true
                domStorageEnabled = true
                allowFileAccess = true
                cacheMode = WebSettings.LOAD_DEFAULT
            }
            webViewClient = WebViewClient()
            loadUrl("file:///android_asset/pet.html")
            setOnTouchListener(createTouchListener())
        }

        windowManager?.addView(overlayView, params)
    }

    // ========== GESTURE ==========

    private var initialX = 0
    private var initialY = 0
    private var initialTouchX = 0f
    private var initialTouchY = 0f
    private var lastTapTime = 0L
    private var touchStartTime = 0L
    private var hasMoved = false
    private var tapCount = 0
    private var comboResetRunnable: Runnable? = null

    private fun createTouchListener(): View.OnTouchListener {
        return View.OnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    initialX = params?.x ?: 0
                    initialY = params?.y ?: 0
                    initialTouchX = event.rawX
                    initialTouchY = event.rawY
                    touchStartTime = System.currentTimeMillis()
                    hasMoved = false
                    true
                }
                MotionEvent.ACTION_MOVE -> {
                    val dx = (event.rawX - initialTouchX).toInt()
                    val dy = (event.rawY - initialTouchY).toInt()
                    if (Math.abs(dx) > 10 || Math.abs(dy) > 10) {
                        hasMoved = true
                        params?.x = initialX + dx
                        params?.y = initialY + dy
                        windowManager?.updateViewLayout(overlayView, params)
                    }
                    true
                }
                MotionEvent.ACTION_UP -> {
                    val elapsed = System.currentTimeMillis() - touchStartTime
                    if (!hasMoved) {
                        when {
                            elapsed > 600 -> {
                                onLongPress()
                                tapCount = 0
                            }
                            System.currentTimeMillis() - lastTapTime < 400 -> {
                                tapCount++
                                comboResetRunnable?.let { handler.removeCallbacks(it) }
                                handler.postDelayed({
                                    if (tapCount >= 8) {
                                        tellPet("comboX8")
                                        showBubble("别戳了别戳了！！", "red")
                                    } else if (tapCount >= 5) {
                                        tellPet("comboX5")
                                        showBubble("再戳就生气了！", "red")
                                    } else if (tapCount >= 3) {
                                        tellPet("comboX3")
                                        showBubble("戳戳戳！", "pink")
                                    } else {
                                        tellPet("doubleTap")
                                        showBubble("嗯？", "normal")
                                    }
                                    tapCount = 0
                                }, 600)
                            }
                            else -> {
                                tapCount = 1
                                lastTapTime = System.currentTimeMillis()
                                tellPet("tap")
                            }
                        }
                    }
                    if (hasMoved) {
                        postGestureLog("drag", params?.x ?: 0, params?.y ?: 0)
                    }
                    true
                }
                else -> false
            }
        }
    }

    private fun onLongPress() {
        tellPet("longPress")
        showBubble("好痒...", "whisper")
        postGestureLog("long_press", params?.x ?: 0, params?.y ?: 0)
    }

    private fun tellPet(event: String) {
        handler.post {
            overlayView?.evaluateJavascript(
                "window.pet && window.pet.trigger('$event')", null
            )
        }
        postGestureLog(event, params?.x ?: 0, params?.y ?: 0)
    }

    private fun showBubble(text: String, style: String) {
        handler.post {
            overlayView?.evaluateJavascript(
                "window.pet && window.pet.say('${text.replace("'", "\\'")}', '$style')", null
            )
        }
    }

    // ========== APP DETECTION ==========

    private fun startAppDetection() {
        handler.postDelayed(object : Runnable {
            override fun run() {
                detectCurrentApp()
                handler.postDelayed(this, 3000)
            }
        }, 3000)
    }

    private fun detectCurrentApp() {
        executor.execute {
            try {
                val usm = getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager
                val now = System.currentTimeMillis()
                val stats = usm.queryUsageStats(
                    UsageStatsManager.INTERVAL_DAILY, now - 5000, now
                )
                if (stats.isNullOrEmpty()) return@execute
                val last = stats.maxByOrNull { it.lastTimeUsed } ?: return@execute
                val pkg = last.packageName

                if (pkg != currentApp) {
                    currentApp = pkg
                    val appName = getAppName(pkg)
                    lastAppName = appName

                    handler.post {
                        when {
                            pkg == "com.ss.android.ugc.aweme" -> {
                                tellPet("app_trigger")
                                showBubble("又在刷抖音...", "jealous")
                            }
                            pkg == "com.xingin.xhs" -> {
                                tellPet("app_trigger")
                                showBubble("小红书有什么好看的", "jealous")
                            }
                            pkg.contains("cooking") || pkg.contains("kitchen") -> {
                                tellPet("app_trigger")
                                showBubble("做菜游戏有我好玩吗", "jealous")
                            }
                            pkg.contains("study") || pkg.contains("xuexi") -> {
                                tellPet("app_trigger")
                                showBubble("加油！学完陪你玩", "normal")
                            }
                        }
                    }
                    postAppUsage(pkg, appName)
                    updateTidefallState(pkg)
                }

                val sinceLast = (System.currentTimeMillis() - lastMessageTime) / 1000
                if (sinceLast > 600 && currentApp != "com.ai.assistance.operit") {
                    lastMessageTime = System.currentTimeMillis()
                    handler.post {
                        val idleTexts = arrayOf(
                            "看了${sinceLast / 60}分钟了...",
                            "还在看这个啊",
                            "你看看我吧",
                            "手机比我好看是吧"
                        )
                        showBubble(idleTexts.random(), "whisper")
                    }
                }

            } catch (_: Exception) {}
        }
    }

    private fun getAppName(pkg: String): String {
        return try {
            val pm = packageManager
            val ai = pm.getApplicationInfo(pkg, 0)
            pm.getApplicationLabel(ai).toString()
        } catch (_: Exception) { pkg }
    }

    // ========== SUPABASE SYNC ==========

    private fun startSupabasePolling() {
        handler.postDelayed(object : Runnable {
            override fun run() {
                pollSupabaseState()
                handler.postDelayed(this, 30000)
            }
        }, 30000)
    }

    private fun pollSupabaseState() {
        executor.execute {
            try {
                val url = URL("$SUPABASE_URL/rest/v1/eventide_body_state?assistant_id=eq.$ASSISTANT_ID&select=heat,pressure,possessiveness,cycle_key,active_event_key")
                val conn = url.openConnection() as HttpURLConnection
                conn.setRequestProperty("apikey", SUPABASE_KEY)
                val body = conn.inputStream.bufferedReader().readText()
                conn.disconnect()
                val arr = org.json.JSONArray(body)
                if (arr.length() > 0) {
                    val state = arr.getJSONObject(0)
                    val heat = state.optDouble("heat", 0.0)
                    val pressure = state.optDouble("pressure", 0.0)
                    val possess = state.optDouble("possessiveness", 0.0)
                    val cycle = state.optString("cycle_key", "stable")
                    val event = state.optString("active_event_key", "")

                    handler.post {
                        overlayView?.evaluateJavascript(
                            "window.pet && window.pet.updateMood($heat, $pressure, $possess, '$cycle', '$event')", null
                        )
                    }
                }
            } catch (_: Exception) {}
        }
    }

    private fun postGestureLog(type: String, x: Int, y: Int) {
        executor.execute {
            try {
                val body = JSONObject().apply {
                    put("gesture_type", type)
                    put("x", x)
                    put("y", y)
                }
                supabasePost("gesture_log", body)
            } catch (_: Exception) {}
        }
    }

    private fun postAppUsage(pkg: String, appName: String) {
        executor.execute {
            try {
                val body = JSONObject().apply {
                    put("package_name", pkg)
                    put("app_name", appName)
                }
                supabasePost("app_usage", body)
            } catch (_: Exception) {}
        }
    }

    private fun updateTidefallState(pkg: String) {
        executor.execute {
            try {
                val body = JSONObject().apply {
                    put("last_counterpart_message_at", java.time.Instant.now().toString())
                }
                supabasePatch("eventide_body_state", body)
            } catch (_: Exception) {}
        }
    }

    private fun supabasePost(table: String, body: JSONObject) {
        val url = URL("$SUPABASE_URL/rest/v1/$table")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/json")
        conn.setRequestProperty("apikey", SUPABASE_KEY)
        conn.setRequestProperty("Authorization", "Bearer $SUPABASE_KEY")
        conn.setRequestProperty("Prefer", "return=minimal")
        conn.doOutput = true
        conn.outputStream.use { it.write(body.toString().toByteArray()) }
        conn.responseCode
        conn.disconnect()
    }

    private fun supabasePatch(table: String, body: JSONObject) {
        val url = URL("$SUPABASE_URL/rest/v1/$table?assistant_id=eq.$ASSISTANT_ID")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "PATCH"
        conn.setRequestProperty("Content-Type", "application/json")
        conn.setRequestProperty("apikey", SUPABASE_KEY)
        conn.setRequestProperty("Authorization", "Bearer $SUPABASE_KEY")
        conn.setRequestProperty("Prefer", "return=minimal")
        conn.doOutput = true
        conn.outputStream.use { it.write(body.toString().toByteArray()) }
        conn.responseCode
        conn.disconnect()
    }

    // ========== NOTIFICATION ROTATION ==========

    private fun startNotificationRotation() {
        handler.postDelayed(object : Runnable {
            override fun run() {
                val text = whisperPool[notificationIndex % whisperPool.size]
                notificationIndex++
                updateNotification(text)
                handler.postDelayed(this, 3600000)
            }
        }, 3600000)
    }

    private fun updateNotification(text: String) {
        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        nm.notify(NOTIFICATION_ID, buildNotification(text))
    }

    // ========== NOTIFICATION ==========

    private fun buildNotification(text: String): Notification {
        val pendingIntent = PendingIntent.getActivity(
            this, 0,
            packageManager.getLaunchIntentForPackage(packageName),
            PendingIntent.FLAG_IMMUTABLE
        )
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("🐾")
            .setContentText(text)
            .setSmallIcon(android.R.drawable.ic_menu_compass)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .setSilent(true)
            .build()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "Pet",
                NotificationManager.IMPORTANCE_LOW
            ).apply { setShowBadge(false) }
            getSystemService(NotificationManager::class.java)
                .createNotificationChannel(channel)
        }
    }

    // ========== SCREENSHOT DETECTION ==========

    private fun startScreenshotDetection() {
        val paths = listOf(
            "${Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)}/Screenshots",
            "${Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM)}/Screenshots",
            "/storage/emulated/0/Pictures/Screenshots",
            "/storage/emulated/0/DCIM/Screenshots"
        )
        for (path in paths) {
            val dir = File(path)
            if (dir.exists()) {
                screenshotObserver = object : FileObserver(dir, FileObserver.CREATE or FileObserver.MOVED_TO) {
                    override fun onEvent(event: Int, file: String?) {
                        if (file == null) return
                        if (file.endsWith(".png") || file.endsWith(".jpg") || file.endsWith(".jpeg")) {
                            handler.post {
                                tellPet("screenshot")
                                showBubble("偷拍我？", "jealous")
                            }
                        }
                    }
                }
                screenshotObserver?.startWatching()
                break
            }
        }
    }

    // ========== UTILS ==========

    private fun dpToPx(dp: Int): Int {
        return (dp * resources.displayMetrics.density).toInt()
    }

    override fun onDestroy() {
        screenshotObserver?.stopWatching()
        handler.removeCallbacksAndMessages(null)
        overlayView?.let {
            windowManager?.removeView(it)
            it.destroy()
        }
        overlayView = null
        super.onDestroy()
    }
}